app/src/main/java/com/celzero/bravedns/service/ConnectionMonitor.kt

```diff
@@ -0,0 +1,241 @@
+    // the value loaded from memory instead of Thread's local cache.
+    @Volatile private var handlerThread: HandlerThread? = null
+    private var serviceHandler: NetworkRequestHandler? = null
+    private val persistentState by inject<PersistentState>()
+
+    // https://kotlinlang.org/api/latest/jvm/stdlib/kotlin.collections/-linked-hash-set/
+    // https://developer.android.com/reference/kotlin/java/util/LinkedHashSet
+    // LinkedHashSet maintains the iteration ordering, which is the order in which elements were
+    // inserted into the set (insertion-order)
+    // CurrentNetworkList first element should be always active network. So need to maintain the
+    // insertion order for this Set. (Using LinkedHashSet).
+    var currentNetworkList: LinkedHashSet<Network> = linkedSetOf()
+    var currentActiveNetwork: Network? = null
+
+    var connectivityManager: ConnectivityManager = context.applicationContext.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
+
+
+    companion object {
+        // add active network as underlying vpn network
+        const val MSG_ADD_ACTIVE_NETWORK = 1
+        // add all available networks as underlying vpn networks
+        const val MSG_ADD_ALL_NETWORKS = 2
+    }
+
+    init {
+        connectivityManager.registerNetworkCallback(networkRequest, this)
+        this.handlerThread = HandlerThread("NetworkRequestHandler.HandlerThread")
+        this.handlerThread?.start()
+        this.serviceHandler = NetworkRequestHandler(handlerThread?.looper, this)
+    }
+
+    interface NetworkListener {
+        fun onNetworkDisconnected()
+        fun onNetworkConnected(networkSet: LinkedHashSet<Network>?)
+    }
+
+    fun removeCallBack() {
+        connectivityManager.unregisterNetworkCallback(this)
+        destroy()
+    }
+
+    private fun destroy() {
+        this.serviceHandler?.removeCallbacksAndMessages(null)
+        this.handlerThread?.quitSafely()
+        this.handlerThread = null
+        this.serviceHandler = null
+    }
+
+    override fun onAvailable(network: Network) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onAvailable")
+        handleNetworkChange()
+    }
+
+    override fun onLost(network: Network) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onLost")
+        handleNetworkChange()
+    }
+
+    override fun onCapabilitiesChanged(network: Network, networkCapabilities: NetworkCapabilities) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onCapabilitiesChanged")
+        handleNetworkChange()
+    }
+
+    override fun onLinkPropertiesChanged(network: Network, linkProperties: LinkProperties) {
+        if(DEBUG) Log.d(LOG_TAG, "ConnectionMonitor - onLinkPropertiesChanged")
+        handleNetworkChange()
+    }
+
+    /**
+     * Called when the preference(isAddAllNetwork) is changed.
+     * if the preference is true, add only active network
+     * else, add all available network.
```

Comment:
Also, `onUserPreferenceChanged` warrants special handling, because the current* and new* networks would be unchanged and yet the underlying network needs to be reset according to user pref. For ex, consider the app is connected to multiple networks, `A` (active) and `B`.  User changes pref to connect to a single network. `processActiveNetwork` would see that the current* and new* networks are basically unchanged, and do nothing. The app would still be connected to multiple networks.

In essence, whenever `onUserPreferenceChanged` is called, the messages should be processed forcefully regardless of the current* and new* networks.

One way I can think of is to set a boolean called `forceUpdate` in the handler `msg` obj to `true`, which makes `processActiveNetwork` and `processAllNetworks` bypass those optimizations and straight up call `onNetworkConnected` and `onNetworkDisconnected` as appropriate, after setting current* networks.

Reply:
Added a variable in the message object to check if the forceUpdate is required. If yes, then the setUnderlyingNetwork is updated regardless of the change.

Reply:
Okay.

Forgot to change the comment on the method? See the first entry in this thread.

Reply:
I missed it. Changed the comment.